src/OutputWriters/fetch_output.jl

```diff
@@ -1,16 +1,28 @@
 using CUDA
 
 using Oceananigans.Fields: AbstractField, compute_at!
+using Oceananigans.ImmersedBoundaries: mask_immersed!
 using Oceananigans.Models.LagrangianParticleTracking: LagrangianParticles
 
 # Needed to support `fetch_output` with `model::Nothing`.
 time(model) = model.clock.time
 time(::Nothing) = nothing
 
-fetch_output(output, model) = output(model)
+# Default fetch_output with mask_immersed = nothing
+fetch_output(output, model) = fetch_output(output, model, nothing)
 
-function fetch_output(field::AbstractField, model)
+function fetch_output(output, model, mask_immersed)
+    fetched = output(model)
+    if fetched isa Field
+        !isnothing(mask_immersed) && mask_immersed!(fetched, mask_immersed)
+        return parent(fetched)
+    end
+    return fetched
+end
+
+function fetch_output(field::AbstractField, model, mask_immersed)
     compute_at!(field, time(model))
+    !isnothing(mask_immersed) && mask_immersed!(field, mask_immersed)
```

Comment:
@glwagner I think this won't work because what happens is that, e.g., if `u` is the output to be saved then `model.velocities.u` is masked and then it includes NaNs at the next step so the whole state NaN...

At least that's what I found happening and I boil it down to this line here masking and then returning the parent of the field. Am I right?

Reply:
Ok that makes sense. But the real problem is how we define masking. Consider: the mask we apply to a field for output _should not affect time-stepping_. If it does, there's a problem. So let's fix that before hacking around it.

The issue is velocity components I think. The "mask" is defined to cover the points that are _on_ the boundary, not just inside it. Thus masking the velocity to 0 is how we ensure impenetrability.

The confusion is therefore: are velocity nodes _on_ the boundary immersed or not?

I think we should probably define this points as _not_ immersed. Therefore to fix the problem we can:

1. Implement a new function that satisfies impenetrability. We can start by doing this just for immersed grids.
2. Define "mask_immersed" as something that only affects cells that are _fully_ immersed. This has no effect on fully centered fields, but for staggered fields this implies a change: cells are only masked if they are not adjacent to active cells (I think the distinction between `inactive_node` and `peripheral_node` can be used for this).

Reply:
Yes, I was thinking of point 2 you make and modified `masked_immersed!` to use `inactive_node` (just to test the hypothesis) but that didn't work either.

Reply:
There might also be a problem with the implicit solver to fix. Our strategy right now is to zero out elements of the matrix that correspond to immersed cells. Unfortunately `0 * NaN = NaN` so any `NaN` in the immersed boundary will be propagated immediately in the whole column

Reply:
Good point.

Can't we just apply the mask to the saved output?

Reply:
Note that in contrast, `false * NaN = 0.0`

Reply:
> Good point.
> 
> Can't we just apply the mask to the saved output?

Yes but it's more complicated, since output can be windowed / sliced. It's possible though.

Reply:
> 

What do you mean by this exactly?

Reply:
> Can't we just apply the mask to the saved output?

What do you mean?

Reply:
> There might also be a problem with the implicit solver to fix. Our strategy right now is to zero out elements of the matrix that correspond to immersed cells. Unfortunately 0 * NaN = NaN so any NaN in the immersed boundary will be propagated immediately in the whole column

The immersed solver can then apply `mask_immersed!(velocities, 0)`.

Reply:
I am referring to the vertically implicit solver. We do not use preconditioners, we use the Thomas algorithm and solve the tridiagonal system directly

Reply:
Oh _that_ "implicit solver".

I guess that is the code here:

https://github.com/CliMA/Oceananigans.jl/blob/a1f4f4e75e51a83cf9dec9e671866cff67de1211/src/ImmersedBoundaries/grid_fitted_immersed_boundaries.jl#L126-L148

Seems like we would need

https://github.com/CliMA/Oceananigans.jl/blob/a1f4f4e75e51a83cf9dec9e671866cff67de1211/src/Solvers/batched_tridiagonal_solver.jl#L179

to return 0 for `immersed_cell`.

To accomplish this without explicit masking we might have to wrap `field` in a `ConditionalOperation`:

https://github.com/CliMA/Oceananigans.jl/blob/a1f4f4e75e51a83cf9dec9e671866cff67de1211/src/TurbulenceClosures/vertically_implicit_diffusion_solver.jl#L197-L199

hmm...

Another possibility is to move `mask_immersed!` from `update_state!`; for example if we masked fields before `ab2_step!` then I think we'd be ok. Not sure if that would affect non-masked output. Hopefully not.

Reply:
> > Can't we just apply the mask to the saved output?
> 
> What do you mean?

I mean only mask the data written in the jld2/nc file, not the actual data in the u, v, w, fields, etc.

Reply:
That would require allocating a temporary array (in the GPU case we already do this on the CPU; however we would then have to mask on the CPU which could slow things down for large fields).

Reply:
But maybe what you're suggesting could be achieved by digging into JLD2's array-saving code. I'm not sure about NetCDF.